### ISSUE TITLE
Allow Base Color and Alpha Map textures to be combined on export

### ISSUE THREAD

**Issue by dsinni** [KhronosGroup/glTF-Blender-IO#343]
Currently, it appears that alpha transparency in textures can only be used effectively when it's included in the `Base Color` texture.

It would be amazing if the exporter was capable of combining the `Base Color` texture with an Alpha Map/Mask on export, possibly via `Add Shader` or `Mix Shader` + `Transparent BSDF` (or whichever makes the most sense).

Example:
![image](https://user-images.githubusercontent.com/4733018/53760592-57755c80-3e91-11e9-900c-3c29e32984c4.png)

This would greatly benefit project workflows where separate Alpha Maps/Masks are used.

Thanks for the consideration and all the great work.

**Comment by donmccurdy** [comment]
Maybe a prerequisite question is, should `Transparent BSDF`, mixed with other nodes, be a supported way to export transparency? Without that the node graph above will not work anyway, and you'd need to connect the textures in a new node before connecting to the `Base Color` socket.

^My preference would be to get the Principled BSDF node's alpha channel reflected better in the Blender preview window, and then we wouldn't need to go after the `Transparent BSDF` node, which does not translate as well to glTF. The Blender devs have mentioned they want this too, but that change is probably more than I can manage. 

In the meantime, I'm not sure what I think about supporting `Transparent BSDF`... we'd have to guess whether it's using 'CLIP' or 'MASK' modes I think. 😕

> This would greatly benefit project workflows where separate Alpha Maps/Masks are used.

Is there a particular situation where separate alpha maps are preferable? I've certainly seen it in e.g. some COLLADA models, but I'm not sure why.

**Comment by dsinni** [comment]
Hi, @donmccurdy.

> Maybe a prerequisite question is, should Transparent BSDF, mixed with other nodes, be a supported way to export transparency?

I'm not partial to `Transparent BSDF` in any way; I was just providing a vague example of what I thought could be a possibility in order to combine the Base Color and Alpha texture on export.

> My preference would be to get the Principled BSDF node's alpha channel reflected better in the Blender preview window

I actually believe there have been improvements here. Last I checked, the Base Color alpha channel was being reflected correctly in the viewport, which wasn't the case until somewhat recently.

> Is there a particular situation where separate alpha maps are preferable? I've certainly seen it in e.g. some COLLADA models, but I'm not sure why.

In any case where the Base Color texture doesn't included alpha, say, when using a tiled base texture and want to map alpha to that texture without creating more faces, similarly to how emissive maps are used.

I've seen multiple instances of texture packs containing alpha maps/masks with their base textures. When trying to use them in Blender, I realized that I couldn't figure out how to export the alpha since they weren't combined.

I hope that makes sense.

**Comment by donmccurdy** [comment]
Ok, I see. In that case we can focus on the alpha map question here and leave `Transparent BSDF` for another thread, it should be totally possible to support separate Alpha Map textures without it.

> In any case where the Base Color texture doesn't included alpha, say, when using a tiled base texture and want to map alpha to that texture without creating more faces, ...

The big caveat here is that if your base color is tiled and alpha is not, or vice versa, merging means baking the tiling and wasting a lot of texture space. Dealing with UV transforms when merging textures is probably out of scope for this exporter, but...

... I would totally support having the exporter be able to merge color and alpha channels of two textures into a single base color texture. That seems analogous to what it needs to do already for occlusion/roughness/metalness, and presumably can rely on mostly the same code.

**Comment by dsinni** [comment]
> The big caveat here is that if your base color is tiled and alpha is not, or vice versa, merging means baking the tiling and wasting a lot of texture space. Dealing with UV transforms when merging textures is probably out of scope for this exporter, but...

Ah, yes, Fair point. I can see that potentially being an issue, especially since they need to share the same UVs/transforms, if I understand correctly.

> ... I would totally support having the exporter be able to merge color and alpha channels of two textures into a single base color texture. That seems analogous to what it needs to do already for occlusion/roughness/metalness, and presumably can rely on mostly the same code.

Cool. Yeah, that was my thinking, to combine them similarly to how the ORM texture is handled. I think this would be really useful in cases where they're provided separately. It's likely a matter of convenience, but a big one at that; having to combine them elsewhere and re-import them is a bit tedious for something that can likely be done automatically within Blender/exporter.

Thanks for your insight, @donmccurdy.

**Comment by dsinni** [comment]
@donmccurdy It appears I may have been wrong about improvements in regard to `Principled BSDF` shader's Alpha preview in Blender viewport, unless there has been a recent regression. I was using the `Background` shader, which previews as expected, but still not the case for `Principled BSDF`. Sorry for the confusion.

**Comment by scurest** [comment]
The exporter appears to have picked up this ability at some point.

Exporting with this setup:

![setup](https://user-images.githubusercontent.com/11024420/73486245-5cc17c80-436a-11ea-91f7-c63a1b1dac47.png)

And reimporting:

![result](https://user-images.githubusercontent.com/11024420/73485989-c9884700-4369-11ea-8eef-1ec7c40d00bc.png)

**Comment by emackey** [comment]
Might be good to have the importer use the new `alpha` input on Principled BSDF, but, that has to wait for 2.79 support to go away.  It's 2020 now so it's just whoever has time to make it happen, I think.

In the meantime I think we can consider this issue resolved.

**Comment by Ali-RS** [comment]
This does not work for me! 

@scurest I tried like in your screenshot but no texture is exported for me. 

My shader setup: 

![Screenshot_2020-03-01_16-46-49](https://user-images.githubusercontent.com/5787647/75626457-f9cb3b80-5bdc-11ea-8aa3-dbe510132c66.jpg)



Can you please verify if it works for you?
Model link : 
[test-alpha-map.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/4271802/test-alpha-map.zip)

Using blender 2.81 
Tested with addon version 1.2.17 and 1.2.33 (installed manually)

Regards

**Comment by Ali-RS** [comment]
Never mind, I solved it. Problem was that alpha map size was different from the color map.

**Comment by kokotron** [comment]
[b36_b41_color_alpha_combine_tests.zip](https://github.com/user-attachments/files/15585402/b36_b41_color_alpha_combine_tests.zip)

Hey, sorry for reviving this, but I noticed some incosistent behaviour between versions. I am currently using Blender 4.1 and the color/alpha texture combination only works if the alpha texture includes an alpha channel, even if you use it's color output for alpha. On the other hands, while testing this more, I noticed that in Blender 3.6 it only works if the alpha texture DOES NOT have an alpha channel - the exact opposite of 4.1. I have included a zip file with .blend and .glb files for both occasions.
